### PR TITLE
feat(pagination): re-importing from AIP-158

### LINKED ISSUE #12: Adopt AIP-0158 Pagination

### PR DESCRIPTION
**Pull request by toumorokoshi**
AEP-158 forked from the general aip.dev, whose guidance diverged a bit from google.aip.dev.

This re-imports those, resulting in the following set of changes:

- switching typescript examples to proto (with placeholder for openapi).
- minor wording fixes.

Fixes #12 

## 🍱 Types of changes

What types of changes does your code introduce to AEP? _Put an `x` in the boxes
that apply_

- [ ] Enhancement
- [ ] [New proposal](https://aep.dev/1#workflow)
- [x] Migrated from google.aip.dev
- [ ] Chore / Quick Fix

## 📋 Your checklist for this pull request

Please review the [AEP Style and Guidance](https://aep.dev/style-guide) for
contributing to this repository.

### General

- [x] Basic [Guidance](https://aep.dev/style-guide#guidance) is met.
- [x] Ensure that your PR
      [references AEPs](https://aep.dev/style-guide#referencing-aeps)
      correctly.
- [x] [My code has been formatted](https://aep.dev/contributing#formatting)
      (usually `prettier -w .`)

### CODE REVIEW

File: aep/general/0158/aep.md.j2 [hunk 1]
```diff
@@ -1,100 +1,103 @@
...
+often grow over time, increasing lookup time as well as the size of the
+responses being sent over the wire. Therefore, it is important that collections
+be paginated.
 
 ## Guidance
 
-Operations returning collections of data **must** provide pagination _at the
-outset_, as it is a [backwards-incompatible change](#backwards-compatibility)
-to add pagination to an existing method.
+RPCs returning collections of data **must** provide pagination _at the outset_,
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:11]:
```suggestion
Methods returning collections of data **must** provide pagination _at the outset_,
```

We've settled on "method" as the proper generic term for referring to things that may be gRPC or OAS (or something else).

File: aep/general/0158/aep.md.j2 [hunk 2]
```diff
@@ -1,100 +1,103 @@
...
+    (google.api.resource_reference) = {
+      child_type: "library.googleapis.com/Book"
+    }];
 
   // The maximum number of books to return. The service may return fewer than
   // this value.
   // If unspecified, at most 50 books will be returned.
   // The maximum value is 1000; values above 1000 will be coerced to 1000.
-  maxPageSize: bigint;
+  int32 page_size = 2;
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:32]:
```suggestion
  int32 max_page_size = 2;
```

Also needs updating elsewhere.  `max_page_size` is a more accurate name because the server may choose to return fewer than this many results (maybe there's an overfetch + filter in memory strategy and so you often end up with 90ish results rather than the full 100). 

Switching the name of this field was actually one of the few specific things JJ (or Luke) suggested we change, because they thought the AIPs had made a mistake.

File: aep/general/0158/aep.md.j2 [hunk 3]
```diff
@@ -1,100 +1,103 @@
...
-      number requested (including zero results), even if not at the end of the
-      collection.
-- Request definitions for collections **should** define a `string page_token`
+  - If the user specifies `page_size` greater than the maximum permitted by the
+    API, the API **should** coerce down to the maximum permitted page size.
+  - If the user specifies a negative value for `page_size`, the API **must**
+    send an `INVALID_ARGUMENT` error.
+  - The API **may** return fewer results than the number requested (including
+    zero results), even if not at the end of the collection.
+- Request messages for collections **should** define a `string page_token`
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:71]:
```suggestion
- Request schemas for collections **should** define a `string page_token`
```

"messages" is proto-specific.  "definitions" works but I think we've used "schemas" elsewhere.

File: aep/general/0158/aep.md.j2 [hunk 4]
```diff
@@ -1,100 +1,103 @@
...
+  - If the user changes the `page_size` in a request for subsequent pages, the
+    service **must** honor the new page size.
+  - The user is expected to keep all other arguments to the RPC the same; if
+    any arguments are different, the API **should** send an `INVALID_ARGUMENT`
+    error.
 - The response **must not** be a streaming response.
-- Services **may** support using page tokens across versions of a service, but
-  are not required to do so.
-- Response definitions for collections **must** define a
+- Response messages for collections **should** define a
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:80]: Why weaken must to should here?

File: aep/general/0158/aep.md.j2 [hunk 5]
```diff
@@ -1,100 +1,103 @@
...
   `string next_page_token` field, providing the user with a page token that may
   be used to retrieve the next page.
-  - The field containing pagination results **should** be the first field
-    specified. It **should** be a repeated field containing a list of resources
-    constituting a single page of results.
+  - The field containing pagination results **should** be the first field in
+    the message and have a field number of `1`. It **should** be a repeated
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:84]: "and have a field number of `1`" is protobuf-specific.

File: aep/general/0158/aep.md.j2 [hunk 6]
```diff
@@ -133,32 +136,21 @@ authorization to the underlying resources, and authorization **must** be
...
+used. It is not necessary to document this behavior.
 
-Services **may** choose to be strongly consistent by approximating the
-"repeatable read" behavior in databases, and returning exactly the records that
-exist at the time that pagination begins.
+**Note:** While a reasonable time may vary between APIs, a good rule of thumb
+is three days.
 
-### Backwards compatibility
+## Backwards compatibility
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:149]: Why the change here?  I think this still belongs under `## Guidance`.

File: aep/general/0158/aep.md.j2 [hunk 7]
```diff
@@ -167,48 +159,17 @@ now is only getting the first 50 (and does not know to advance pagination).
...
 
 **Warning:** This also entails that, in addition to presenting the pagination
 fields, they **must** be _actually implemented_ with a non-infinite default
 value. Implementing an in-memory version (which might fetch everything then
 paginate) is reasonable for initially-small collections.
 
-## Implementation
+[list]: ./list
 
-Page tokens **should** be versioned independently of the public API, so that
```
Comment (rofrankel) [inline aep/general/0158/aep.md.j2:181]: Why remove this section?  Do you think it's not useful guidance?

**Review by rofrankel — CHANGES_REQUESTED** [review]
A lot of the changes in this PR take generic wording and make it proto-specific (outside the protobuf tab).  Other changes revert the improvement from `page_size` to `max_page_size`. 

I commented on the first instance of each that I noticed, but not on all instances.

File: aep/general/0158/aep.md.j2 (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0158/aep.md.j2:11]: I thought it was operations? e.g. https://aep.dev/151. @mkistler - I forgot where we landed, but we're still mixing that nomenclature in AEPs.

File: aep/general/0158/aep.md.j2 (hunk 7 above)
Comment (mkistler) [reply aep/general/0158/aep.md.j2:181]: It looks like this section was removed (or maybe never present?) in the Google AIP.

**Review by mkistler — APPROVED** [review]
Looks good. 👍

File: aep/general/0158/aep.md.j2 (hunk 7 above)
Comment (rofrankel) [reply aep/general/0158/aep.md.j2:181]: Discussed live: we'll drop this section but file a ticket to keep track of potentially adding it back.

**Review by rofrankel — APPROVED** [review]
